### PR TITLE
Internalize UDQ keywords

### PR DESCRIPTION
**Pull request by joakim-hove**
This a *very* initial start of UDQ support; currently just some parsing and keywords.

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

File: tests/parser/UDQTests.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+#include <opm/parser/eclipse/EclipseState/Runspec.hpp>
+#include <opm/parser/eclipse/Parser/Parser.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/UDQ.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/UDQRecord.hpp>
+
+using namespace Opm;
+
+BOOST_AUTO_TEST_CASE(KEYWORDS) {
+    const std::string input = R"(
+    RUNSPEC
```
Comment (bska) [inline tests/parser/UDQTests.cpp]: Do note that if you write it like this, then the leading blanks on each line become part of the string.  In other words, this is strictly speaking not valid ECL input (which requires that all keywords start in column 1).

You can verify by running the example below
```C++
#include <iostream>
#include <string>

int main() {
    const auto s = std::string { R"(
        Hi there
    )" };

    std::cout << s;
} 
```

Comment (joakim-hove) [reply tests/parser/UDQTests.cpp]: Thank you - will fix!

**Comment by joakim-hove** [comment]
jenkins build this please

File: CMakeLists_files.cmake [hunk 2]
```diff
@@ -447,6 +450,9 @@ if(ENABLE_ECL_INPUT)
        opm/parser/eclipse/EclipseState/IOConfig/IOConfig.hpp
        opm/parser/eclipse/EclipseState/checkDeck.hpp
        opm/parser/eclipse/EclipseState/Runspec.hpp
+       opm/parser/eclipse/EclipseState/UDQCOnfig.hpp
```
Comment (bska) [inline CMakeLists_files.cmake]: This should probably say "Config" (lower-case "o").

File: src/opm/parser/eclipse/EclipseState/Schedule/UDQRecord.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+
+        void assertKeyword(const std::string& keyword) {
+            const std::string valid_start = "CFGRSWAB";
+
+            if (valid_start.find(keyword[0]) == std::string::npos)
+                throw std::invalid_argument("Leading character must be in set: " + valid_start);
+
+            if (!keyword[1] == 'U')
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/UDQRecord.cpp]: Did you mean
```C++
keyword[1] != 'U'
```
here (or `!(keyword[1] == 'U')` for that matter)?

**Comment by joakim-hove** [comment]
Ehhh yes - this was embarrassingly full of errors. In progress....

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
OK - I would like to merge this now; it is a *far shot* away from usable UDQ support - but it is a start.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
I notice that you have started using `noexcept`, do you intend to do this throughout? It is not enforced at compile time, but do your compilers give warnings that help avoiding unnecessary exceptions?

**Comment by joakim-hove** [comment]
> I notice that you have started using noexcept, do you intend to do this throughout?

Ehh - to be 100% honest I have just copied the PR's from @bska; my understand of the `noexcept` is quite shallow.

**Comment by atgeirr** [comment]
> Ehh - to be 100% honest I have just copied the PR's from @bska; my understand of the noexcept is quite shallow.

I'll discuss it with @bska then... :-)

**Comment by atgeirr** [comment]
As this has been reviewed by @bska and Jenkins is happy, it should be merged.

**Comment by joakim-hove** [comment]
> As this has been reviewed by @bska 

I am not 100% certain @bska himself feels he has reviewed it - but it is not very controversial anyway.

**Comment by atgeirr** [comment]
I apologize if I was too quick! If so, I'll take corrective steps as necessary.

**Comment by joakim-hove** [comment]
> I'll take corrective steps as necessary.

I don't think that is necessary; of course personally I think the code was ready. As of yet it has no users, so ... ?

**Pull request merged**